### PR TITLE
[FSDP] Add fast path for `NO_SHARD` `clip_grad_norm_()`

### PR DESCRIPTION
**Pull request by awgu**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#89137 [FSDP] Add fast path for `NO_SHARD` `clip_grad_norm_()`**

### CODE REVIEW

File: test/distributed/fsdp/test_fsdp_clip_grad_norm.py [hunk 1]
```diff
@@ -209,6 +209,34 @@ def _test_ddp_parity(
                 self.assertEqual(n1, n2)
                 self.assertEqual(p1, p2)
 
+        if offload_params:
+            # TODO: Gradient computation on CPU and GPU differ slightly causing
+            # drift unrelated to `clip_grad_norm_()`.
+            # https://github.com/pytorch/pytorch/issues/89133
+            return
```
Comment (rohan-varma) [inline test/distributed/fsdp/test_fsdp_clip_grad_norm.py:216]: A bit hacky, but would it help if we just ran DDP on CPU for this case if having this coverage is important?

File: test/distributed/fsdp/test_fsdp_clip_grad_norm.py [hunk 2]
```diff
@@ -209,6 +209,34 @@ def _test_ddp_parity(
...
+
+        # Run a few more iterations
+        # TODO: We cannot run too many iterations, or else there is drift:
+        # https://github.com/pytorch/pytorch/issues/89136
+        for _ in range(3):
+            ddp_optim.zero_grad(set_to_none=True)
+            fsdp_optim.zero_grad(set_to_none=True)
```
Comment (rohan-varma) [inline test/distributed/fsdp/test_fsdp_clip_grad_norm.py:223]: nit: are we interested in testing regular zero_grad as well?

File: torch/distributed/fsdp/fully_sharded_data_parallel.py [hunk 3]
```diff
@@ -1161,10 +1161,19 @@ def clip_grad_norm_(
             self._streams["unshard"],
             self._streams["pre_unshard"],
         )
+        # Check for an early return if every FSDP instance uses `NO_SHARD`
```
Comment (rohan-varma) [inline torch/distributed/fsdp/fully_sharded_data_parallel.py:1164]: nit maybe change: Check for an early return if every FSDP instance uses `NO_SHARD` -> "If every FSDP instance uses `NO_SHARD`, early return by calling into the nn.utils helper"?

**Review by rohan-varma — APPROVED** [review]
LGTM, thanks for tracking the follow up issues around improved testing!

File: test/distributed/fsdp/test_fsdp_clip_grad_norm.py (hunk 1 above)
Comment (awgu) [reply test/distributed/fsdp/test_fsdp_clip_grad_norm.py:216]: Does DDP work on CPU with NCCL backend?

**Pull request closed without merging**